cdq-analysis/pkg/devfile.go

```diff
@@ -179,63 +179,86 @@ func ValidateDevfile(log logr.Logger, URL string, token string) (shouldIgnoreDev
+	} else if len(imageComp) > 1 {
+		found := false
+		for _, component := range imageComp {
+			err = validateImageComponentDockerfile(log, component, devfileLocation, token)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("failed to validate the Dockerfile from the Image Component %s in the devfile", component.Name))
 				return shouldIgnoreDevfile, nil, err
 			}
+
+			if _, ok := deployCompMap[component.Name]; ok {
+				found = true
+				break
+			}
 		}
+		if !found {
+			err = fmt.Errorf("found more than one image components, but no deploy command associated with any being defined in the devfile from %s", devfileLocation)
+			log.Error(err, "failed to validate devfile")
+			return shouldIgnoreDevfile, nil, err
+		}
+	} else if len(imageComp) == 1 {
 		// TODO: if only one image component, should return a warning that no apply command being defined
+		component := imageComp[0]
+		err = validateImageComponentDockerfile(log, component, devfileLocation, token)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to validate the Dockerfile from the Image Component %s in the devfile", component.Name))
+			return shouldIgnoreDevfile, nil, err
+		}
 	}
 
 	return shouldIgnoreDevfile, devfileBytes, nil
 }
 
+// validateImageComponentDockerfile validates the given image component dockerfile for a devfile location
+func validateImageComponentDockerfile(log logr.Logger, component v1alpha2.Component, devfileLocation, token string) (err error) {
+	if component.Image != nil && component.Image.Dockerfile != nil && component.Image.Dockerfile.DockerfileSrc.Uri != "" {
+		dockerfileURI := component.Image.Dockerfile.DockerfileSrc.Uri
+		absoluteDockerfileURI := strings.HasPrefix(dockerfileURI, "http://") || strings.HasPrefix(dockerfileURI, "https://")
+		absoluteDevfileLocation := strings.HasPrefix(devfileLocation, "http://") || strings.HasPrefix(devfileLocation, "https://")
+
+		if absoluteDockerfileURI {
+			// absolute Dockerfile uri
+			log.Info(fmt.Sprintf("Checking if the Dockerfile location %s is reachable", dockerfileURI))
+			_, err = CurlEndpoint(dockerfileURI, token)
+		} else {
+			if !absoluteDevfileLocation {
+				// local devfile src with relative Dockerfile uri
+				dockerfileURI = path.Join(path.Dir(devfileLocation), dockerfileURI)
+				log.Info(fmt.Sprintf("Checking if the Dockerfile location %s is reachable", dockerfileURI))
+				err = parserUtil.ValidateFile(dockerfileURI)
+			} else {
+				// remote devfile src with relative Dockerfile uri
+				var u *url.URL
+				u, err = url.Parse(devfileLocation)
+				if err != nil {
+					log.Error(err, fmt.Sprintf("failed to parse URL from %s", devfileLocation))
+					return fmt.Errorf(fmt.Sprintf("failed to parse URL from %s", devfileLocation))
+				}
+				u.Path = path.Join(path.Dir(u.Path), dockerfileURI)
+				dockerfileURI = u.String()
+				log.Info(fmt.Sprintf("Checking if the Dockerfile location %s is reachable", dockerfileURI))
+				_, err = CurlEndpoint(dockerfileURI, token)
```

Comment:
I get that we want to make this function common across controllers but since we clone in the CDQ, I'm wondering if we should re-factor this logic or introduce a new function in the future to search in the cloned repo location rather than make an unnecessary curl call.

Reply:
I wonder how else would you verify a Dockerfile URI. Even if its cloned, you have to do a GET..

Reply:
I was going by the scenario comment `remote devfile src with relative Dockerfile uri`. If it's relative to the remote devfile, it should be in the cloned repo?

Reply:
> If it's relative to the remote devfile, it should be in the cloned repo. We can use the util.IsExist to determine if it's in the fs

Ah, ok i see your point! But I guess if I want to use the cloned path in here, it is going to change the function signature a lot. Let me see if it can be done cleanly

Reply:
yes, we can handle it in another issue if it's too complex of a change

Reply:
Actually as I was making the change to use the local cloned path for validating the devfile, I noticed that in CDQ we always return `DevfileURL` in the status. Build Services and others use this direct devfile URL. Hence I feel, it is better if we validate the Devfile using the URL in this scenario which would be most of the time. 🤔

Reply:
Plus, it also ensures us that the URL we are returning is accessible via the network